### ISSUE TITLE
Avoid exposing low-level wrappers to the consumer

### ISSUE THREAD

**Issue by glopesdev** [opentk/opentk#1008]
OpenTK has been developed in the past as a high-level interface to audio, graphics, input and windowing. Although it featured fast, automatically generated low-level bindings to OpenGL and other libraries, the goal was to move away from low-level wrappers such as the Tao bindings and towards more C# / .NET friendly interfaces.

Therefore, low-level wrappers in OpenTK have typically been regarded as an implementation detail of a higher level architecture. This is also convenient because it gives the development team freedom to replace the implementation of this low-level wrapper as requirements evolve.

The current OpenAL wrapper included as part of OpenTK now seems to have dramatically changed direction. The low-level implementation details are bubbling to a high-level interface which is exposing pointers, unsafe methods and interfaces, and essentially exposing low-level implementation details alongside its high-level interface.

Why was this decision made and what were the perceived advantages behind it? By keeping the high-level interface short and minimalist, I believe it is simpler both for users to enjoy OpenTK, and for maintainers to evolve the architecture, without having to juggle the underlying legacy.

**Comment by PJB3005** [comment]
OpenAL and OpenGL are high level interfaces?

Personally I would prefer if lower-level APIs were exposed decently. Of course there should still be higher level APIs like with say GL context creation. Personally I feel like the GLFW binding strikes this balance very well (GLFW exposed directly if you need it, but all functionality available high level).

**Comment by glopesdev** [comment]
My reference to high or low level interfaces was referring to the C# API and not OpenGL/OpenAL itself. In this context, I consider the P/Invoke layer with unsafe pointers to be the "low-level" interface, and the set of static method classes such as `GL` to be the "high-level" interface.

The difference between them is that in a high-level interface the user does not have to be exposed to p/invoke, low-level callbacks and pointers, or C-style error handling, but rather the concepts have been "translated" into the host language style (e.g. in this case C#).

In most successful foreign language interfaces I have developed and used in the past, the translation layer is so pleasant and performant that there is hardly ever any need to use the low-level entry points.

Even more importantly, avoiding exposing the low-level native bindings allows maintainers to change the way the bindings are done, without breaking user code. For example, we could have just switched between p/invokes, delegates and/or fody without breaking any user code.

If we expose the low-level bindings, we are affirming our commitment to a particular implementation, which as we have experienced these past few years is more likely to change than we would have hoped (and not likely to stabilize any time soon imo).

I don't see any reason why we can't converge on a set of high-level (or idiomatic) APIs that would be both performant and conformant to C# style (e.g. using Span<T> and other efficient but safe constructs).

**Comment by PJB3005** [comment]
> In this context, I consider the P/Invoke layer with unsafe pointers to be the "low-level" interface

You're mixing things up that should be mixed up here. The P/Invoke layer and the pointer API are two separate things and I'll elaborate down below.

> Even more importantly, avoiding exposing the low-level native bindings allows maintainers to change the way the bindings are done, without breaking user code. For example, we could have just switched between p/invokes, delegates and/or fody without breaking any user code.
>
> If we expose the low-level bindings, we are affirming our commitment to a particular implementation, which as we have experienced these past few years is more likely to change than we would have hoped (and not likely to stabilize any time soon imo).

We're never exposing the P/Invoke layer itself. The closest you're getting to it is `GetProcAddress()` and the hook for it, but those are extremely necessary in many use cases, and also standard in other OpenGL bindings.

> In most successful foreign language interfaces I have developed and used in the past, the translation layer is so pleasant and performant that there is hardly ever any need to use the low-level entry points.

The keyword here is **hardly**, not **never**. People have exotic use cases all the time and this is no exception. Maybe you're binding to a C library or converting an example for said library to C#. Using pointers directly could very well be the easiest solution. If we did not expose `GL.BufferSubData(..., int size, IntPtr data)` directly and instead exposed `GL.BufferSubData<T>(..., Span<T> data) where T : unmanaged` you're forced to jump through non-obvious hoops to access the *actual* OpenGL API you need.

> I don't see any reason why we can't converge on a set of high-level (or idiomatic) APIs that would be both performant and conformant to C# style (e.g. using Span and other efficient but safe constructs).

This is quite literally impossible with some of the constructs in OpenGL's API. For example `glShaderSource` takes a `GLchar **`. You cannot sanely map that in a safe and performant way because you'd need some API like `Span<Span<byte>>` which isn't possible. With "performant" here I mean "no allocations or conversions of data done by OpenTK".

Exposing these APIs does not cost much from our end (seriously, it's some overloads, people are gonna use the `T[]` or `Span<T>` overload in VS before they use the `T*` one), and they are quite literally the "official" OpenGL API.

**Comment by NogginBops** [comment]
I totally agree with @PJB3005 on this. This is a wrapper for opengl, it's not some library that builds ontop of it. If opengl takes a pointer as an argument then we should have that overload in c#. Then, to make it more usable in c# we provide convenience overloads for c# types that can be interpreted as pointers. If opengl does error handling in some way (`glGetError`) then that should be reflected in c#. The same goes for openal. Opentk 3.x seems to not being able to decide between a high level object oriented wrapper or a low level wrapper (for openal). This is fixed in opentk 4 as the whole high-level stuff is removed. opentk 4 will just be a openal wrapper nothing more complicated. It will expose an api with pointers and convenience overloads for those. Nothing more.

Also of note is that much of the api exposed isn't really implementation dependent. I would almost bet that you don't need to change a single opengl call then going to opentk 4 (except context creation, but that's arguably not opengl). Same goes with openal (though some namespaces changed and unless you are using the "high level" features which are gone in opentk 4)

**Comment by glopesdev** [comment]
> You're mixing things up here. The P/Invoke layer and the pointer API are two separate things and I'll elaborate down below.

I explained myself poorly, and I do apologize. I agree that safe/unsafe and p/invoke are two separate layers, but in most idiomatic wrapper designs I have encountered, the decisions around them usually go hand in hand.

Let me elaborate on the patterns I generally use to pass buffered data around in C# by discussing two imaginary call signatures (not from actual OpenTK code):

```c#
public static void UpdateData<T>(int handle, T[] data) where T : struct;
```

I use this pattern to take care of all meaningful buffered data in native managed memory that could be moved around before the advent of `Span<T>`: blittable data. Any managed array that is of a blittable value type will have a continuous memory layout which can be converted directly to a pointer type using P/invoke without copying or converting the array. The p/invoke/Marshal API just needs to fix it and it's done.

More recently, the runtime has even been extended to support generic blittable types, so this case is only getting better for wrapper implementers: https://github.com/dotnet/runtime/pull/103

Furthermore, with the new `Unsafe` class, it's easier than ever to do more flexible unsafe memory conversions without relying on complex conversion code so this case is becoming really easy.

```c#
public static void UpdateData(int handle, IntPtr data, int size);
```

This one I use to take care of all pointer-based data living in native unmanaged memory land. Its completely safe, works for both heap and stack memory, and even if you have an unsafe pointer lying around, you can always cast it back to an `IntPtr`. This includes the case you presented with `GLchar **`.

It does place the responsibility of handling pointers squarely back to the user, but in my opinion this is actually an advantage. I really think that for all the _hard_ cases where you do need low-level pointer manipulations for performance reasons, the programmer really should know what they are doing. Having an API that does not expose unsafe pointers avoids encouraging reckless behavior. Also as I discussed before, it gives flexibility to the wrapper maintainers to adopt new techniques available in the .NET runtime as they become better understood, and safely made available (such as `Span<T>`). Having the `IntPtr` overloads keeps all the flexibility that the expert needs when absolutely necessary.

The only really new player in the game is `Span<T>` which does introduce enough performance and safe language features that I completely agree makes it deserve an overload for it.

The reason for avoiding unsafe is that from my experience it is always a slippery slope. Once you expose it in public APIs, you will almost always invariably tend to abuse it, even for cases that don't require it. Keeping it out of wrappers entirely personally helped me in the past to develop an appreciation for how many times you actually don't need it at all (with no loss of performance), resulting in API designs which are easier to understand.

To summarize and clarify, I think that we should:

1. absolutely include `IntPtr` and `Span<T>` APIs;
2. carefully consider the need to include unsafe native pointer APIs on a _case-by-case_ basis;

Can you point me to examples of number 2.? As a user who has been constantly shuffling data between managed/unmanaged world using OpenTK in the past years, I have personally never felt the need to use pointers, neither for consistency, nor performance.

However, I am very open to changing my mind, as there might be some really tricky cases which I may have missed.

**Comment by glopesdev** [comment]
I also want to clarify that one of the main reasons not to expose a pointer to the managed world is the intrinsic ambiguity on the meaning of a pointer. A pointer in native land can be:

- an input argument (by ref)
- an output argument (by ref)
- an array (of unknown length)

This ambiguity is what makes pointer APIs dangerous and confusing, even in native land. C# has built-in ways to explicitly make the intended semantics very clear, which do not exist in C. Because it supports overloads even on P/invoke calls, you can even have multiple of the above resolve to the same call, if two of the possible semantics are supported.

These can be used without sacrificing performance, and you can still use IntPtr for all the edge cases. _Unsafe_ pointer types in C# exist for local pointer manipulation, when necessary, and are generally not recommended in public APIs.

Now that the runtime even supports generic blittable types directly in the p/invoke layer, I'm finding it even harder to justify.

**Comment by NogginBops** [comment]
Idk what need to be justified. `int*` is better than `IntPtr`. When opengl wants a `GLchar**` we provide a `byte**` overload as well as `IntPtr`, `string`, `T[]`, and whatever else. If you don't have a pointer you can't use the pointer overload. If you have a pointer, fine just use it, no casting needed. The fact that the API exposes pointers is nothing weird, because it's expected for opengl. 

> Unsafe pointer types in C# exist for local pointer manipulation, when necessary, and are generally not recommended in public APIs.

This is not at all true. Pointer types existed in .net long before stack memory manipulation was a focus. Pointers in .net was for building fast code that could interop with the native part of the system. There is nothing that says libraries shouldn't expose APIs containing pointers.

> Now that the runtime even supports generic blittable types directly in the p/invoke layer, I'm finding it even harder to justify.

We can use this as we want to support .net framework which doesn't do blittable generics.

**Comment by glopesdev** [comment]
Nothing needs to be justified. These are suggestions based on my experience designing and maintaining several low-level wrappers for C#. Take them or leave them as you prefer, of course.

**Comment by glopesdev** [comment]
I agree we are looking at two different approaches: design an object model to expose OpenGL/OpenAL to the c# world, or a low-level p/invoke API.

The old Tao.Bindings were an example of a P/invoke. They were painful to use. OpenTK felt really nice because it felt like C#. In the few places where you needed performance, it was always there.

Can you give me a concrete example at least of when does this approach fail? Just so it stays in the record for historical reasons when you close the issue, so we can understand in the future the motivations.

**Comment by NogginBops** [comment]
It's not that a `IntPtr` overload isn't going to work. It's just going to be more work for people to use. When using opengl it is not uncommon to interface with native libraries or write code where you really want to just pass a pointer to a gl function. 

I personally don't see how it's a problem exposing a pointer overload. You can't use it if you don't already have a pointer, so it's not going to cause any problem for people who aren't using pointers. And for people who are using pointers, it's a lot easier/cleaner. In my eyes it's a win-win.

And if you are using opengl it's kind of expected that you know how pointers work. It's just the nature of the API and the complexity level of opengl.

I'm not sure I understand your concern. There is no part of the API that forces you to use `unsafe`, you can do everything fine with just managed stuff. And if you need the raw opengl, it's there.

> If we expose the low-level bindings, we are affirming our commitment to a particular implementation...
> Even more importantly, avoiding exposing the low-level native bindings allows maintainers to change the way the bindings are done...

The statements above are simply not true, because we are exposing opengl exactly as specified by the standard. So there are no implementation details of the bindings exposed there.

**Comment by glopesdev** [comment]
Can you provide a link to a code sample where pointers were heavily used and constantly passed around? All of my C# development experience has been spent doing interop between native/managed code, and I found the most maintainable/performant designs leave the pointer manipulation in the native world.

Usually my experience has been that whenever there is such a heavy use of pointer manipulations that such APIs would be even be convenient its a red flag that design should be reviewed.

In fact, I often find the opposite: using unsafe everywhere is much more annoying than just doing a cast, as you have to be enclosing every scope, or every function, with the unsafe keyword, and even mark your entire project as unsafe. I usually avoid it as much as possible as it tends to spread and invites both reckless behaviour and requires the highest level of elevation/trust in your runtime to even run.

Again, this may just be my personal experience and those of my colleagues, so I think it would be really useful to understand a satisfying design that makes such a heavy use of unsafe that would justify a blanket pass to provide every single OpenGL/OpenAL overload with pointer support.

**Comment by NogginBops** [comment]
I think this issue boils down to not wanting `unsafe` overloads in the API. And I think this is just a case of; If you don't want it, don't use it. You can disable unsafe code from your solution, that way you won't be able to use these overloads.

Many people expect pointers to be available when dealing with APIs like this. Having only a `IntPtr` API for these people is taking away type safety for these use cases and that is not generally useful.

**Comment by PJB3005** [comment]
I had to re-read your first comment *multiple* times because it felt like a disconnected tangent at first, but I finally managed to get at what you mean and uh, I can kind of pin down where you're mistaken, to put it bluntly.

> To summarize and clarify, I think that we should:
> * absolutely include IntPtr and Span<T> APIs;
> * carefully consider the need to include unsafe native pointer APIs on a case-by-case basis;

> Its completely safe, works for both heap and stack memory, and even if you have an unsafe pointer lying around, you can always cast it back to an `IntPtr`. This includes the case you presented with `GLchar **`.

**You seem to be mixing up unsafe and `unsafe`.** This API absolutely is not safe. Just because you don't need the C# `unsafe` keyword doesn't mean that it's "safe" to do these pointer operations. Somebody can EASILY pass `IntPtr.Zero` to that method, without enabling unsafe code in their project, and get an access violation/segfault. 

In this case, the practical difference between `IntPtr` and `void*` (which you advocate against) is that:
1. `void*` needs `unsafe` keyword. (this is a good thing)
2. `void*` is slightly easier to work with as a, you know, pointer.
3. `IntPtr` is ambiguous between "pointer sized int" and "poorly typed pointer". `void*` is unambiguous.
4. `void*` cannot be used from VB. (who cares)

**Now, I add those up, and `IntPtr` just has no benefits (unless you really like VB, I guess?)**

> Now that the runtime even supports generic blittable types directly in the p/invoke layer, I'm finding it even harder to justify.

You don't need blittable generics in P/Invoke signatures to be able to do this. Just turn it into a `void*` by casting some pointers and pass that to OpenGL. 

> I also want to clarify that one of the main reasons not to expose a pointer to the managed world is the intrinsic ambiguity on the meaning of a pointer. 

It's low level for a reason and a necessary sacrifice. Don't need to lecture us on why pointers are dangerous. 

> These can be used without sacrificing performance, and you can still use IntPtr for all the edge cases. 

Yes, obviously? Was this ever cast into doubt? Obviously we'll use `ref`, `in` and `out` where appropriate in the "safe" GL API. 

> Unsafe pointer types in C# exist for local pointer manipulation, when necessary, and are generally not recommended in public APIs.

Why would not use them in method signatures? VB has to use them?

> I agree we are looking at two different approaches: design an object model to expose OpenGL/OpenAL to the c# world, or a low-level p/invoke API.

What? Where are you getting this idea from?? OpenTK is **not** an OOP OpenGL/OpenAL API. Not at all. It's a raw OpenGL binding with some higher-level convenience bindings for .NET (like taking in arrays instead of raw pointers). OpenTK is **absolutely** closer to "low-level P/Invoke API" and it *always* has been as far as I know.

> The old Tao.Bindings were an example of a P/invoke. They were painful to use. OpenTK felt really nice because it felt like C#. In the few places where you needed performance, it was always there.

So why can't we be both, like OpenTK 3... already is?

> Can you give me a concrete example at least of when does this approach fail? Just so it stays in the record for historical reasons when you close the issue, so we can understand in the future the motivations.

What approach? Having an *entirely separate library* as escape hatch? I feel that "entirely separate library" speaks for itself. 

> Can you provide a link to a code sample where pointers were heavily used and constantly passed around?

In my game I wrote `Span<T>` wrappers on top of `GL.BufferSubData` for OpenTK 3. This would not have been possible if only the `T[]` overload were exposed.

> and I found the most maintainable/performant designs leave the pointer manipulation in the native world.

In my experience the less native code I have to modify the better, but whatever.

> Usually my experience has been that whenever there is such a heavy use of pointer manipulations that such APIs would be even be convenient its a red flag that design should be reviewed.

These APIs will always have safe managed overloads. I don't see the issue here.

> In fact, I often find the opposite: using unsafe everywhere is much more annoying than just doing a cast, as you have to be enclosing every scope, or every function, with the unsafe keyword, and even mark your entire project as unsafe. I usually avoid it as much as possible as it tends to spread and invites both reckless behaviour and requires the highest level of elevation/trust in your runtime to even run.

That is literally the entire point. It's annoying and easy to find because it is *unsafe* and it should be easily auditable.

> Again, this may just be my personal experience and those of my colleagues, so I think it would be really useful to understand a satisfying design that makes such a heavy use of unsafe that would justify a blanket pass to provide every single OpenGL/OpenAL overload with pointer support.

Like Nogging said, why do you care? Just don't enable unsafe and don't use them...

**Comment by glopesdev** [comment]
> Many people expect pointers to be available when dealing with APIs like this.

Do you have a reference for this? I really have the exact opposite impression from many years of using c#/.NET and I’m genuinely wondering whether I might have missed some fundamental change in code style and C# patterns in the last couple of years?

Also, this is not just about the use of `unsafe` in function overloads. This might have changed now, but when I first started trying to use the new code base I found several instances of unsafe structures and classes containing unsafe-only public members, such as NativeWindow, like: https://github.com/opentk/opentk/blob/master/src/OpenToolkit.Windowing.Desktop/NativeWindow.cs

This in my opinion is where the slippery slope begins, as it becomes then not a matter of giving more choice to developers, but actively promoting that the only way of accessing certain fields is unsafe. This is for me the biggest red flag.

If there are really compelling reasons to have low-level overloads in order to have compatibility with external user defined types, that is one thing, but to actively promote unsafe types is indicative of a very different design approach which I don’t understand.

**Comment by glopesdev** [comment]
> Like Nogging said, why do you care? Just don't enable unsafe and don't use them...

If I want to access the WindowState of the native window, I literally don’t have the choice.

I apologize if I said something to offend, I will step back now, as I don’t have the personal bandwidth to take this right now.

**Comment by NogginBops** [comment]
@glopesdev  NOW I GET IT!! With the code sample it becomes really clear what you mean! I agree that the API shouldn't expose structures with unsafe members like that. I was thinking you where talking about OpenGL functions exclusively. But now that I see what you are referencing I agree that it should probably be changed if possible. 

Unsafe fields force the user to use unsafe, and that is not good. It should be up to the user if they want to use unsafe or not. So here I definitely agree with you that it should be changed to not be an unsafe field.

**Comment by PJB3005** [comment]
> Do you have a reference for this? I really have the exact opposite impression from many years of using c#/.NET and I’m genuinely wondering whether I might have missed some fundamental change in code style and C# patterns in the last couple of years?

I don't have "evidence" but I know I can use at least myself and Noggin as an anecdote. Also C# *has* been gaining a lot of nice low level & pointer features semi-recently so...?

> This might have changed now, but when I first started trying to use the new code base I found several instances of unsafe structures and classes containing unsafe-only public members, such as NativeWindow, like: https://github.com/opentk/opentk/blob/master/src/OpenToolkit.Windowing.Desktop/NativeWindow.cs 

This is a problem how? The alternative to `WindowPtr` is not exposing GLFW at all (if you say that `NativeWindow` shouldn't be locked to GLFW, that's understandable but a *completely* different debate). The GLFW API *is* unsafe. We're exposing it in the hope that it is useful to developers who have the need for it.

**Obviously we will still provide a complete, safe and managed binding of all of GLFW's APIs**, so you only really ever need this for stuff like passing it to other C libraries. That said I don't believe this is currently the case but it is the intent when everything is done.

> If I want to access the WindowState of the native window, I literally don’t have the choice.

`unsafe` on a method/property declaration doesn't mean you need to be in an unsafe context to access it. it's just a shortcut for wrapping the entire function body in `unsafe { }`. Are you mixing this up with Rust where this *is* the case?

**Comment by varon** [comment]
Design decision were made on this. Thanks for the great discussion. Closing.
